### PR TITLE
Preview plugin for displaying PDFs in browser with option to extend

### PR DESCRIPTION
**Pull request by nicwells**
Implements https://github.com/BlueBrain/nexus/issues/2566

Preview plugin will display when there is a distribution. It will list any PDFs in the distribution and provide the option to preview, download or copy the URI. It can be extended to allow other resources besides pdf's - csv's for example. Also includes the options available in the data access plugin to copy a uri and download the file

PDF_Viewer_demo

### CODE REVIEW

File: src/shared/components/Preview/Preview.tsx [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+            nexus={nexus}
+            url={previewAsset.url}
+            previewDivRef={previewDivRef}
+          />
+        ) : (
+          <></>
+        )
+      ) : (
+        <></>
+      )}
```
Comment (Stafeeva) [inline src/shared/components/Preview/Preview.tsx:199]: Can we simplify it like:
```
{previewAsset && previewAsset.encodingFormat === 'application/pdf' && (
    <PDFViewer
      resource={resource}
      nexus={nexus}
      url={previewAsset.url}
      previewDivRef={previewDivRef}
    />
)}
```

File: src/shared/containers/ResourceViewContainer.tsx [hunk 2]
```diff
@@ -309,18 +310,22 @@ const ResourceViewContainer: React.FunctionComponent<{
...
+                    <div style={{ marginTop: 15 }}>
+                      <Alert
+                        message="You don't have access to edit the resource. You can nonetheless see the resource metadata below."
+                        type="info"
+                      />
+                      {resource.distribution ? (
+                        <Preview nexus={nexus} resource={resource} />
+                      ) : (
+                        <></>
+                      )}
```
Comment (Stafeeva) [inline src/shared/containers/ResourceViewContainer.tsx:328]: Ans same here, we can just do:
```
{resource.distribution && (<Preview nexus={nexus} resource={resource} />)}
```

File: src/shared/containers/ResourceViewContainer.tsx [hunk 3]
```diff
@@ -342,23 +347,24 @@ const ResourceViewContainer: React.FunctionComponent<{
...
                   typeof resource['@type'] === 'string' &&
                   nonEditableResourceTypes.includes(resource['@type']) && (
-                    <p>
-                      <Alert
-                        message="Please not for the time being this resource is not editable. For further information please contact the administrator."
-                        type="info"
-                      />
-                    </p>
+                    <Alert
+                      message="Please not for the time being this resource is not editable. For further information please contact the administrator."
```
Comment (Stafeeva) [inline src/shared/containers/ResourceViewContainer.tsx:359]:
```suggestion
                      message="Please note for the time being this resource is not editable. For further information please contact the administrator."
```

File: src/shared/containers/ResourceViewContainer.tsx [hunk 4]
```diff
@@ -342,23 +347,24 @@ const ResourceViewContainer: React.FunctionComponent<{
...
+                      type="info"
+                    />
                   )}
+                {resource.distribution ? (
+                  <Preview nexus={nexus} resource={resource} />
+                ) : (
+                  <></>
+                )}
```
Comment (Stafeeva) [inline src/shared/containers/ResourceViewContainer.tsx:367]: Same suggestion, just makes it tidier and easier to read
```
{resource.distribution && (<Preview nexus={nexus} resource={resource} />)}
```

**Comment by smitfire** [comment]
One deepscan issue, reviewing code further atm

File: yarn.lock [hunk 5]
```diff
@@ -3007,6 +3007,11 @@
   resolved "https://registry.yarnpkg.com/@types/parse-json/-/parse-json-4.0.0.tgz#2f8bb441434d163b35fb8ffdccd7138927ffb8c0"
   integrity sha512-//oorEZjL6sbPcKUaCdIGlIUeH26mgzimjBB77G6XRgnDl/L5wOnpyBGRe/Mmf5CVW3PwEBE1NjiMZ/ssFh4wA==
 
+"@types/pdfjs-dist@*":
```
Comment (smitfire) [inline yarn.lock:3010]: we shouldn't add lock files to PRs :)

File: src/shared/components/Preview/Preview.tsx [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+  nexus: NexusClient;
+}> = ({ resource, nexus }) => {
+  const notification = useNotification();
+  const [previewAsset, setPreviewAsset] = React.useState<any | undefined>();
+  const previewDivRef = React.useRef<HTMLDivElement>(null);
+  const [orgLabel, projectLabel] = parseProjectUrl(resource._project);
+
+  /* close preview when anywhere outside the preview is clicked */
+  React.useEffect(() => {
+    document.addEventListener('mousedown', e => {
```
Comment (Stafeeva) [inline src/shared/components/Preview/Preview.tsx:43]: is it possible to use antd Modal window here? it has all the styles and built in functionality like click outside to close etc

File: yarn.lock (hunk 5 above)
Comment (nicwells) [reply yarn.lock:3010]: lock files need to be in there no?

File: src/shared/components/Preview/Preview.tsx (hunk 6 above)
Comment (nicwells) [reply src/shared/components/Preview/Preview.tsx:43]: I did try with the antd modal but it was introducing styling challenges. Have left as is and fixed the clean up issue picked up by deepscan

Comment (Stafeeva) [reply src/shared/components/Preview/Preview.tsx:43]: 👍

**Review by Stafeeva — APPROVED** [review]

**Review by dhaneshnm — APPROVED** [review]

**Review by Stafeeva — APPROVED** [review]

**Pull request merged**